src/coreclr/jit/flowgraph.cpp

```diff
@@ -3116,6 +3116,53 @@ BasicBlock* Compiler::fgLastBBInMainFunction()
     return fgLastBB;
 }
 
+//------------------------------------------------------------------------------
+// fgGetDomSpeculatively: Try determine a more accurate dominator than cached bbIDom
+//
+// Arguments:
+//    block - Basic block to get a dominator for
+//
+// Return Value:
+//    Basic block that dominates this block
+//
+BasicBlock* Compiler::fgGetDomSpeculatively(const BasicBlock* block)
+{
+    assert(fgDomsComputed);
+    int         reachablePreds    = 0;
+    BasicBlock* lastReachablePred = nullptr;
+
+    // Check if we have unreachable preds
+    for (const flowList* predEdge : block->PredEdges())
+    {
+        BasicBlock* predBlock = predEdge->getBlock();
+        if (predBlock == block)
+        {
+            // can a block have a predEdge to itself?
+            return block->bbIDom;
+        }
+
+        // We check pred's count of InEdges - it's quite conservative.
+        // We, probably, could use fgReachable(fgFirstBb, pred) here to detect unreachable preds
+        if (predBlock->countOfInEdges() > 0)
+        {
+            lastReachablePred = predBlock;
+            reachablePreds++;
```

Comment:
Nit: you don't need to count reachable preds here. If `lastReachablePred` is non-null, then there is more than one, the search fails, and you can set it to null and break.

Then below just check `lastReachablePred` is null or not.